genetic_algorithm/travelling_salesman_problem.py

```diff
@@ -0,0 +1,343 @@
+"Given a list of cities and the distances between each pair of cities, what is the
+ shortest possible route that visits each city exactly once and returns to the origin
+ city?"
+
+https://en.wikipedia.org/wiki/Genetic_algorithm
+https://en.wikipedia.org/wiki/Travelling_salesman_problem
+
+Author: Clark
+"""
+
+import copy
+import random
+
+cities = {
+    0: [0, 0],
+    1: [0, 5],
+    2: [3, 8],
+    3: [8, 10],
+    4: [12, 8],
+    5: [12, 4],
+    6: [8, 0],
+    7: [6, 2],
+}
+
+
+def main(
+    cities: dict[int, list[int]],
+    population_size: int,
+    iterations_num: int,
+    crossover_probability: float,
+    mutation_probability: float,
+) -> tuple[list[int], float]:
+    """
+    Genetic algorithm main function
+    >>> main(cities=cities,population_size=100,iterations_num=100,
+    ... crossover_probability=0.6,mutation_probability=0.2)
+    ([0, 1, 2, 3, 4, 5, 6, 7, 0], 37.909778143828696)
+    >>> main(cities={0: [0, 0], 1: [2, 2]},population_size=10,iterations_num=10,
+    ... crossover_probability=0.6,mutation_probability=0.2)
+    ([0, 1, 0], 5.656854249492381)
+    >>> main(cities={},population_size=10,iterations_num=10,
+    ... crossover_probability=0.6,mutation_probability=0.2)
+    Traceback (most recent call last):
+      ...
+    IndexError: list assignment index out of range
+    """
+    best_path: list[int] = []
+    best_distance = float("inf")
+
+    chromosomes, cities_list = init(cities, population_size)
+    fitness_matrix, best_path, best_distance = fitness(
+        cities, chromosomes, best_path, best_distance
+    )
+    for _ in range(iterations_num):
+        """
+        Uncomment to choose another selection operator
+        Only one of the two selection operators can be uncommented at the same time.
+        """
+        # chromosomes = chose_ts(fitness_matrix, chromosomes, population_size)
+        chromosomes = chose_rws(fitness_matrix, chromosomes, population_size)
+        for x in range(int(population_size / 2)):  # Population crossover
+            chromosomes[x], chromosomes[x + int(population_size / 2)] = crossing(
+                chromosomes[x],
+                chromosomes[x + int(population_size / 2)],
+                crossover_probability,
+                cities_list,
+            )
+        for x in range(population_size):  # Population variation
+            chromosomes[x] = mutate(chromosomes[x], mutation_probability)
+
+        fitness_matrix, best_path, best_distance = fitness(
+            cities, chromosomes, best_path, best_distance
+        )
+
+    return best_path, best_distance
+
+
+def distance(city1: list[int], city2: list[int]) -> float:
+    """
+    Calculate the distance between two coordinate points
+    >>> distance([0, 0], [3, 4] )
+    5.0
+    >>> distance([0, 0], [-3, 4] )
+    5.0
+    >>> distance([0, 0], [-3, -4] )
+    5.0
+    """
+    return (((city1[0] - city2[0]) ** 2) + ((city1[1] - city2[1]) ** 2)) ** 0.5
+
+
+def init(
```

Comment:
```suggestion
def init(cities: dict[int, list[int]], population_size: int) -> tuple[list[list[int]], list[int]]:
    """
    Initialization generates initial population
    >>> init(cities={0: [0, 0], 1: [2, 2]}, population_size=2)
    ([[0, 1, 0], [0, 1, 0]], [1])
    >>> init(cities={0: [0, 0], 1: [2, 2]}, population_size=0)
    ([], [1])
    >>> init(cities={}, population_size=2)
    ([], [])
    """
    chromosomes = []
    cities_list = list(cities.keys())
    if cities_list:  # Check if cities_list is not empty before deleting an element
        del cities_list[0]
    for _ in range(population_size):
        chromosome = []
        chromosome.append(0)  # Add starting point
        chromosome.extend(random.sample(cities_list, len(cities_list)))
        chromosome.append(0)  # Add end point
        chromosomes.append(chromosome)
    return chromosomes, cities_list

```

Reply:
Sorry, I don't understand why it needs to be modified like this